### PR TITLE
Add coverage to packages beyond contracts

### PR DESCRIPTION
**Pull request by rajivpo**
**Description**
Added the ability to surface test coverage metrics for all relevant packages. This is the first of hopefully a few PRs I want to submit to improve test coverage across various packages (probably starting with `core-utils` as it's the easiest) and figured having the ability to surface where coverage is lacking in an automated fashion would be helpful for development.

I've also seen repos hook up coverage numbers to their homepages, and if this is a desired feature, I'm happy to investigate. There's also probably an opportunity to ensure that coverage isn't decreasing as a pre-commit hook, and am happy to spend time implementing that would be helpful.

**Additional context**

**Metadata**

### CODE REVIEW

**Comment by tynes** [comment]
Hey @rajivpo thanks for the PR! Do you mind explaining a bit about how `nyc` is used? Perhaps a blurb in the top level `README`. It looks like this PR adds the ability to run `nyc` against each of the different packages - it would be nice to have its output show up in the PR or something like that. I believe we use codecov already for the contracts, it looks like there is an integration: https://github.com/istanbuljs/nyc/blob/master/docs/setup-codecov.md

**Comment by tynes** [comment]
Do you also mind squashing your commits into a single commit?

File: .github/workflows/ts-packages.yml [hunk 1]
```diff
@@ -60,45 +60,8 @@ jobs:
         env:
           CC_SECRET: ${{ secrets.CC_SECRET }}
 
-  # A hack that allows running a job only if a specific directory changed.
```
Comment (rajivpo) [inline .github/workflows/ts-packages.yml:63]: I figured i could remove this entire block now that there's coverage across all ts-packages, but if there's a good reason to keep this, happy to add back in.

**Comment by rajivpo** [comment]
@tynes does this suffice?

**Comment by rajivpo** [comment]
Should have a bit more time this week - will continue adding to coverage to packages

**Comment by elenadimitrova** [comment]
Thanks for the contribution! In line with what @tynes said above, can we also add the Codecov report run on the respective package readme, same way we do for contracts https://github.com/ethereum-optimism/optimism/tree/develop/packages/contracts 
Also we need to add checks for the branch coverage to CI for each package and set it to not drop below what we currently have. I think improving coverage after that can be done in separate PRs to make this and subsequent work easier to review.

**Comment by elenadimitrova** [comment]
@rajivpo heya, just looking at the last commit and seeing that codecov is now _just_ on the main readme, as opposed to what we agreed above which was to add the Codecov report run on each of the respective package readme. Basically I personally think that coverage per package is more meaningful than overall monorepo coverage. Sorry if my comment above was what confused you!

**Comment by rajivpo** [comment]
oh okok yes this makes a lot more sense, I think this can be done with flags via codecov, but will need to double check

**Comment by rajivpo** [comment]
Given how messy this is, I'm going to close this out as a much cleaner version of this is in development in #1412

**Pull request closed without merging**